From 5833dfebd12b5cea7233ec9059f1e7ffae204632 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Tue, 29 Jun 2021 18:19:54 +0200
Subject: [PATCH 119/179] memory: omap-gpmc: irq_pipeline: enable pipelined
 interrupt control

Switching the demultiplexer IRQ to out-of-band mode ensures the
pipelined control flow is followed as expected for all demultiplexed
events via:

generic_pipeline_irq()
   -> <demux_flow_handler>
           -> handle_oob_irq()
                -> decoding_irq_handler() [e.g. gpmc_handle_irq()]
		      for_each_pending_chained_irq() {
                           generic_handle_irq()
                              -> <flow_handler>
                                     -> handle_oob_irq()
		      }

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 drivers/memory/omap-gpmc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/memory/omap-gpmc.c b/drivers/memory/omap-gpmc.c
index f80c2ea39..ebcda04da 100644
--- a/drivers/memory/omap-gpmc.c
+++ b/drivers/memory/omap-gpmc.c
@@ -1405,6 +1405,7 @@ static int gpmc_setup_irq(struct gpmc_device *gpmc)
 	gpmc->irq_chip.irq_mask = gpmc_irq_mask;
 	gpmc->irq_chip.irq_unmask = gpmc_irq_unmask;
 	gpmc->irq_chip.irq_set_type = gpmc_irq_set_type;
+	gpmc->irq_chip.flags = IRQCHIP_PIPELINE_SAFE;
 
 	gpmc_irq_domain = irq_domain_add_linear(gpmc->dev->of_node,
 						gpmc->nirqs,
@@ -1415,7 +1416,7 @@ static int gpmc_setup_irq(struct gpmc_device *gpmc)
 		return -ENODEV;
 	}
 
-	rc = request_irq(gpmc->irq, gpmc_handle_irq, 0, "gpmc", gpmc);
+	rc = request_irq(gpmc->irq, gpmc_handle_irq, IRQF_OOB, "gpmc", gpmc);
 	if (rc) {
 		dev_err(gpmc->dev, "failed to request irq %d: %d\n",
 			gpmc->irq, rc);
-- 
2.38.1

